excel.go

```diff
@@ -0,0 +1,235 @@
+package main
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/xuri/excelize/v2"
+)
+
+type PasswordRow struct {
+	Password string
+	Row      int
+}
+
+func toSheetCoord(coord int) int {
+	return coord + 1
+}
+
+func getHeaderToXCoord(headerRow []string) map[string]int {
+	headerToXCoord := make(map[string]int, len(headerRow))
+	for i, cell := range headerRow {
+		headerToXCoord[cell] = i
+	}
+	return headerToXCoord
+}
+
+func getMACFinUsers(f *excelize.File, input *Input) (map[string]string, error) {
+	rows, err := f.GetRows(input.SheetName)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting rows from %s in %s: %s", input.SheetName, input.Filename, err)
+	}
+
+	users := make(map[string]string)
+
+	headerToXCoord := getHeaderToXCoord(rows[0])
+	usernameXCoord := headerToXCoord[input.UsernameHeader]
+	passwordXCoord := headerToXCoord[input.PasswordHeader]
+	rowOffset := input.RowOffset
+
+	for _, row := range rows[rowOffset:] {
+		// check if row is empty
+		if row == nil {
+			continue
+		}
+		users[row[usernameXCoord]] = row[passwordXCoord]
+	}
+
+	return users, nil
+}
+
+func getManagedUsers(f *excelize.File, input *Input) (map[string]PasswordRow, error) {
+	automatedSheet := input.AutomatedSheetName
+	rows, err := f.GetRows(automatedSheet)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting rows from %s in %s: %s", automatedSheet, input.Filename, err)
+	}
+
+	rowOffset := input.RowOffset
+	colUser := input.AutomatedSheetColNameToIndex[ColUser]
+	colPassword := input.AutomatedSheetColNameToIndex[ColPassword]
+	userToPasswordRow := make(map[string]PasswordRow)
+
+	for i, row := range rows[rowOffset:] {
+		userToPasswordRow[row[colUser]] = PasswordRow{
+			Password: row[colPassword],
+			Row:      i,
+		}
+	}
+
+	return userToPasswordRow, nil
+}
+
+// Sync PasswordManager usernames with MACFin users
+func syncPasswordManagerUsersToMACFinUsers(f *excelize.File, input *Input) error {
+	macFinUsersToPasswords, err := getMACFinUsers(f, input)
+	if err != nil {
+		return err
+	}
+
+	userToPasswordRow, err := getManagedUsers(f, input)
+	if err != nil {
+		return err
+	}
+
+	rowOffset := input.RowOffset
+	numRows := len(userToPasswordRow) + rowOffset
+	automatedSheet := input.AutomatedSheetName
+	numCols := len(input.AutomatedSheetColNameToIndex)
+
+	// add new MACFin users to automatedSheet
+	for macFinUser, password := range macFinUsersToPasswords {
+		if _, ok := userToPasswordRow[macFinUser]; !ok {
+			values := []string{macFinUser, password, password, "Rotate Now"}
```

Comment:
Here you are assuming the order of the columns in the automated sheet regardless of the `input`.

You could do

```
values := make([]string, 4)
values[input.AutomatedSheetColNameToIndex[ColUser]] = macFinUser
// etc
```

Reply:
I made `values` a map literal:
```
			values := map[int]string{
				input.AutomatedSheetColNameToIndex[ColUser]:      macFinUser,
				input.AutomatedSheetColNameToIndex[ColPassword]:  password,
				input.AutomatedSheetColNameToIndex[ColPrevious]:  password,
				input.AutomatedSheetColNameToIndex[ColTimestamp]: "Rotate Now",
			}
```